steel/src/history/mod.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright 2024 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+    beacon, BlockHeaderCommit, Commitment, CommitmentVersion, ComposeInput, EvmBlockHeader,
+};
+use alloy_primitives::{Sealed, B256, U256};
+use beacon::{BeaconCommit, GeneralizedBeaconCommit, STATE_ROOT_LEAF_INDEX};
+use beacon_roots::BeaconRootsContract;
+use serde::{Deserialize, Serialize};
+mod beacon_roots;
+
+/// Input committing a previous block hash to the corresponding Beacon Chain block root.
+pub type HistoryInput<H> = ComposeInput<H, HistoryCommit>;
+
+/// Verifiable proof that an execution block hash is included in the past of a specific beacon block
+/// on the Ethereum blockchain.
+///
+/// This struct encapsulates the necessary data to prove that a given execution block is part of the
+/// canonical chain according to the Beacon Chain.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct HistoryCommit {
```

Comment:
How could a guest constrain the block it runs it's queries against (tied to the `evm_commit`)? As it stands I don't see any way for the guest to e.g. check the block number or hash such that it could either place a constraint on it, or commit it to the journal to be checked in the application contract.

Reply:
In the guest, the `EthEvmEnv` gives you access to the (EVM query) header. You can do things like:
```rust
// Read the input from the guest environment.
let input: EthEvmInput = env::read();
let evm_env = input.into_env();

let block_hash = evm_env.header().seal();
let block_number = evm_env.header().number;
...
```

This header is exactly what will be verified by a `BlockHeaderCommit`. Do you think this is enough for your use case?

Reply:
Ah! That's perfect, thank you